### PR TITLE
JDK-8298118: split-if optimization causes empty loop to temporarily have more than one phi

### PR DESCRIPTION
**Pull request by dafedafe**
## Issue
C2 crashes with an assertion failure while running this very simple test:
```java
static void test(int a) {
    int I;
    for (int j = 6; j < 129; ++j) {
        a >>= 37388;
        a += 1001;
    }
    a >>>= -8;
    for (i = 0; i < 8; ++i) {
        iArrFld[i] = I;
    }
    for (int j = i; j < 7; ) {
        x = a;
    }
}
```
The failing assertion makes sure that only one phi exists when removing empty counted loops (in `IdealLoopTree::do_remove_empty_loop`), i.e. the one of the induction variable.

https://github.com/openjdk/jdk/blob/836198a4009c4a3f10a76dc8734e4792bb2509ba/src/hotspot/share/opto/loopTransform.cpp#L3623-L3629

---
## Cause
The test code produces an empty counted loop with 2 phis: nodes 398 (induction variable) and 495.
image
Phi 495 is dead and should be removed (in fact it is set to be removed in one of the next IGVN iterations but the assert happens before the next iteration).

We get to this situation from here (while running `PhaseIdealLoop::split_if_with_blocks`):
image
_Split-if_ "moves" node 396 before phi 401 (it actually creates new nodes 495 and 496) and additionally one of 495's inputs gets replaced by a constant (singleton).
image

Then dead nodes are removed starting from 396 (396, 401, 395). To avoid killing the new node (495) a use to a temporary node is added during the iterative removal run (the reason for this is explained by @vnkozlov [here](https://github.com/openjdk/jdk/pull/3336#discussion_r610877083)):
image

So, we end up with this graph:
image

It is worth mentioning that the code envisages the removal of the new node (495) if dead, but only in the next iteration (it gets added to the _worklist_).
https://github.com/openjdk/jdk/blob/836198a4009c4a3f10a76dc8734e4792bb2509ba/src/hotspot/share/opto/phaseX.cpp#L1493-L1495

Later on in the same iteration, while running `IdealLoopTree::iteration_split`, the loop body gets "cleaned"
https://github.com/openjdk/jdk/blob/836198a4009c4a3f10a76dc8734e4792bb2509ba/src/hotspot/share/opto/loopTransform.cpp#L4040-L4041
So, the dead phi node (495) is removed from `IdealLoopTree::_body` but it still remains linked to the _CountedLoop_ node (397).
The body is now empty and is therefore scheduled to be removed in `IdealLoopTree::do_remove_empty_loop` but the assert checking for the number of phis attached to the _CountedLoop_ node (397) finds 2 phis and breaks.

---
## Further consideration
The situation that caused this assert to fail is very specific (split-if, counted loop, empty body...) but we can't rule out that another constellation won't produce the same issue.

---
## Possible solutions
I can think of three possible fixes for this issue:

**1.** relaxing the check for the assert so that only phis with no uses are taken into account (the one currently implemented with this PR)
**2.** removing the dead phi node right after replacing
https://github.com/openjdk/jdk/blob/836198a4009c4a3f10a76dc8734e4792bb2509ba/src/hotspot/share/opto/loopopts.cpp#L1139-L1141
```cpp
// Remove possibly dead phi node immediately
if (phi != NULL && phi->outcnt() == 0) {
  _igvn.remove_dead_node(phi);
}
```
**3.** removing the dead phi node when removing it from the loop body in `IdealLoopTree::DCE_loop_body`
```cpp
void IdealLoopTree::DCE_loop_body() {
  for (uint i = 0; i < _body.size(); i++) {
    Node *node = _body.at(i);
    if (node->outcnt() == 0) {
      _body.map(i, _body.pop());
      _phase->_igvn.remove_dead_node(node);
      i--; // Ensure we revisit the updated index.
    }
  }
}
```

**...what do you think?**

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298118](https://bugs.openjdk.org/browse/JDK-8298118): split-if optimization causes empty loop to temporarily have more than one phi

### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [d9a60c6d](https://git.openjdk.org/jdk/pull/12148/files/d9a60c6d0626c7a0a0099f3118e7d4ce64ced25f)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [8ab90bf5](https://git.openjdk.org/jdk/pull/12148/files/8ab90bf533fb9d134c5375f39063a35c28d222d2)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12148/head:pull/12148` \
`$ git checkout pull/12148`

Update a local copy of the PR: \
`$ git checkout pull/12148` \
`$ git pull https://git.openjdk.org/jdk pull/12148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12148`

View PR using the GUI difftool: \
`$ git pr show -t 12148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12148.diff">https://git.openjdk.org/jdk/pull/12148.diff</a>

</details>

### CODE REVIEW

**Review by rwestrel — APPROVED** [review]
Looks good to me. I ran into a similar issue when working on an unrelated change and fixed it the same way.

**Review by TobiHartmann — APPROVED** [review]
Nice analysis and summary! Solution 1) looks also most reasonable to me.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -3622,7 +3622,7 @@ bool IdealLoopTree::do_remove_empty_loop(PhaseIdealLoop *phase) {
   Node* iv = NULL;
   for (DUIterator_Fast imax, i = cl->fast_outs(imax); i < imax; i++) {
     Node* n = cl->fast_out(i);
-    if (n->Opcode() == Op_Phi) {
+    if ((n->Opcode() == Op_Phi) && (n->outcnt() > 0)) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:3625]: May be add comment to explain additional check.

Comment (dafedafe) [reply src/hotspot/share/opto/loopTransform.cpp:3625]: @vnkozlov thanks for the suggestion: I've updated the existing comment instead (I hope it is clear enough).

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by dafedafe** [comment]
@TobiHartmann @rwestrel @vnkozlov thanks a lot for your reviews!

**Pull request closed without merging**